### PR TITLE
[SPARK-48792][SQL] Fix regression  for INSERT with partial column list to a table with char/varchar

### PR DESCRIPTION
**Pull request by yaooqinn**
### What changes were proposed in this pull request?

#41262 introduced a regression by applying literals with char/varchar type in query output for table insertions, see

https://github.com/apache/spark/pull/41262/files#diff-6e331e8f1c67b5920fb46263b6e582ec6e6a253ee45543559c9692a72a1a40ecR187-R188

This causes bugs 

```java
24/07/03 16:29:01 ERROR Executor: Exception in task 0.0 in stage 0.0 (TID 0)
org.apache.spark.SparkException: [INTERNAL_ERROR] Unsupported data type VarcharType(64). SQLSTATE: XX000
	at org.apache.spark.SparkException$.internalError(SparkException.scala:92)
	at org.apache.spark.SparkException$.internalError(SparkException.scala:96)
``` 

```java
org.apache.spark.SparkUnsupportedOperationException: VarcharType(64) is not supported yet.
	at org.apache.spark.sql.errors.QueryExecutionErrors$.dataTypeUnsupportedYetError(QueryExecutionErrors.scala:993)
	at org.apache.spark.sql.execution.datasources.orc.OrcSerializer.newConverter(OrcSerializer.scala:209)
	at org.apache.spark.sql.execution.datasources.orc.OrcSerializer.$anonfun$converters$2(OrcSerializer.scala:35)
	at scala.collection.immutable.List.map(List.scala:247)
```

### Why are the changes needed?

Bugfix

### Does this PR introduce _any_ user-facing change?

no
### How was this patch tested?

new tests
### Was this patch authored or co-authored using generative AI tooling?
no

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TableOutputResolver.scala [hunk 1]
```diff
@@ -246,22 +242,25 @@ object TableOutputResolver extends SQLConfHelper with Logging {
           case a: Alias => a.withName(expectedName)
           case other => other
         }
-        (matchedCol.dataType, expectedCol.dataType) match {
+        val replacedExpectedCol = expectedCol.withDataType {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TableOutputResolver.scala:245]:
```suggestion
        val actualExpectedCol = expectedCol.withDataType {
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TableOutputResolver.scala [hunk 2]
```diff
@@ -288,11 +287,13 @@ object TableOutputResolver extends SQLConfHelper with Logging {
   private def resolveColumnsByPosition(
       tableName: String,
       inputCols: Seq[NamedExpression],
-      expectedCols: Seq[Attribute],
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TableOutputResolver.scala:291]: the previous name looks more correct

File: sql/core/src/test/scala/org/apache/spark/sql/CharVarcharTestSuite.scala [hunk 3]
```diff
@@ -81,7 +81,7 @@ trait CharVarcharTestSuite extends QueryTest with SQLTestUtils {
           checkPlainResult(spark.table("t"), typ, v)
         }
         sql("INSERT OVERWRITE t VALUES ('1', null)")
-        checkPlainResult(spark.table("t"), typ, null)
+         (spark.table("t"), typ, null)
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/CharVarcharTestSuite.scala:84]: why is this change?

Comment (yaooqinn) [reply sql/core/src/test/scala/org/apache/spark/sql/CharVarcharTestSuite.scala:84]: by accident

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TableOutputResolver.scala (hunk 2 above)
Comment (yaooqinn) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TableOutputResolver.scala]: Thought about reduce LOC before, addressed

**Review by cloud-fan — APPROVED** [review]

**Comment by yaooqinn** [comment]
Merged to master, thank you for the review @cloud-fan

**Pull request closed without merging**